File: java_hummus/src/main/java/com/github/klee0kai/hummus/collections/weaklist/WeakList.java
package com.github.klee0kai.hummus.collections.weaklist;

import com.github.klee0kai.hummus.model.IProvide;
import com.github.klee0kai.hummus.model.RefProvide;

import java.util.Collection;
import java.util.List;

public class WeakList<T> extends RefList<T> {

    public WeakList() {

    }

    public WeakList(Collection<? extends T> c) {
        this();
        addAll(c);
    }


    @Override
    IProvide<T> wrapRef(T val) {
        return RefProvide.weak(val);
    }

    @Override
    RefList<T> createNew(List<T> list) {
        return new WeakList<>(list);
    }
}

File: kotlin_hummus/src/main/kotlin/com/github/klee0kai/hummus/collections/weaklist/WeakList.kt
package com.github.klee0kai.hummus.collections.weaklist

import com.github.klee0kai.hummus.model.IProvide
import com.github.klee0kai.hummus.model.RefProvide

open class WeakList<T> constructor() : RefList<T>() {

    constructor(list: Iterable<T?>) : this() {
        addAll(list)
    }

    override fun wrapRef(it: T?): IProvide<T?> {
        return RefProvide.weak(it)
    }

    override fun createNew(list: List<T?>): RefList<T> {
        return WeakList(list)
    }
}